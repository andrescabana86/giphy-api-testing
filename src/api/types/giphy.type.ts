export type TResult = {
  id: string
  title: string
  url: string
}

export type TResponse = {
  data: TGif[]
  meta: {
    msg: string
    response_id: string
    status: number
  }
  pagination: {
    count: number
    offset: number
    total_count: number
  }
}

export type TGif = {
  bitly_gif_url: string
  bitly_url: string
  content_url: string
  embed_url: string
  id: string
  images: { original: TOriginalImage }
  import_datetime: string
  is_sticker: number
  rating: string
  slug: string
  source: string
  source_post_url: string
  source_tld: string
  title: string
  trending_datetime: string
  type: string
  url: string
  user: TUser
  username: string
}

export type TOriginalImage = {
  frames: string
  hash: string
  height: string
  mp4: string
  mp4_size: string
  size: string
  url: string
  webp: string
  webp_size: string
  width: string
}

export type TUser = {
  avatar_url: string
  banner_image: string
  banner_url: string
  description: string
  display_name: string
  instagram_url: string
  is_verified: boolean
  profile_url: string
  username: string
  website_url: string
}
