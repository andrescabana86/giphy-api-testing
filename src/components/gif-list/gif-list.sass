.gif-list
  display: grid
  grid-template-columns: repeat(auto-fit, min(96px, 128px))
  grid-auto-columns: 1fr
  grid-auto-rows: 1fr
  grid-gap: 5px
  width: 600px
  min-width: 320px
  margin: 0 auto
  justify-content: center