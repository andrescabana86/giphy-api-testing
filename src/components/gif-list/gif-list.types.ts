import { TResult, TResultTrending } from '@api/types/giphy.types'

export type TGifListProps = {
  list: (TResult | TResultTrending)[]
  loading: boolean
}
