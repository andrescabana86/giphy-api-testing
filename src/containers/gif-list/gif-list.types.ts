import { TResult, TResultTrending } from '@api/types/giphy.types'

export type TGifListContainerProps<T> = {
  children(container: T): JSX.Element
}

export type TGifListContainerState = {
  favorites: (TResult | TResultTrending)[]
  list: (TResult | TResultTrending)[]
  loading: boolean
}
