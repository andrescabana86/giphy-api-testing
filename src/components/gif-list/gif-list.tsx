import React, { useContext } from 'react'
import { GifPreview } from '@components/gif-preview/gif-preview'
import { FavoritesGifsContext } from '@containers/gif-list/gif-list'
import { TGifListProps } from './gif-list.types'
import './gif-list.sass'

export const GifList = React.memo((props: TGifListProps) => {
  const { list } = props
  const FavoritesContext = useContext(FavoritesGifsContext)
  const { favorites } = FavoritesContext
  return (
    <section className="gif-list" data-testid="gif-list">
      {list.map((gif) => (
        <GifPreview
          id={gif.id}
          key={gif.id}
          selected={!!favorites.find((favorite) => favorite.id === gif.id)}
          url={gif.url}
        />
      ))}
    </section>
  )
})
