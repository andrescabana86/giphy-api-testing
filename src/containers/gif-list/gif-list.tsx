import React, { createContext, PureComponent } from 'react'
import { getTrending } from '@api/trending/trending'
import { searchGiphy } from '@api/search/search'
import { TResult, TResultTrending } from '@api/types/giphy.types'
import { TGifListContainerProps, TGifListContainerState } from './gif-list.types'

export const FavoritesGifsContext = createContext<{
  favorites: (TResult | TResultTrending)[]
  saveFavorite?(id: string): void
  removeFavorite?(id: string): void
}>({ favorites: [] })

export class GifListContainer extends PureComponent<
  TGifListContainerProps<GifListContainer>,
  TGifListContainerState
> {
  state: TGifListContainerState = {
    favorites: [],
    list: [],
    loading: true,
  }

  componentDidMount() {
    this.getTrendingList()
  }

  onSearch = (query: string) => {
    this.setState({ loading: true })
    searchGiphy(query).then((results) => {
      this.setState({ loading: false, list: results })
    })
  }

  getTrendingList = () => {
    this.setState({ loading: true })
    getTrending().then((results) => {
      this.setState({ list: results, loading: false })
    })
  }

  onClear = () => {
    this.getTrendingList()
  }

  saveFavorite = (id: string) => {
    const { list, favorites } = this.state
    const favorite = list.find((item) => item.id === id)
    if (favorite) {
      this.setState({
        favorites: [...favorites, favorite],
      })
    }
  }

  removeFavorite = (id: string) => {
    const { favorites } = this.state
    const updatedList = favorites.filter((item) => item.id !== id)
    this.setState({
      favorites: [...updatedList],
    })
  }

  render() {
    const { favorites } = this.state

    return (
      <FavoritesGifsContext.Provider
        value={{ favorites, saveFavorite: this.saveFavorite, removeFavorite: this.removeFavorite }}
      >
        {this.props.children(this)}
      </FavoritesGifsContext.Provider>
    )
  }
}
