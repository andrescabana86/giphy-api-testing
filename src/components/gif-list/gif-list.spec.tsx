import React from 'react'
import { TResult, TResultTrending } from '@api/types/giphy.types'
import { render, cleanup } from '@testing-library/react'
import { GifList } from './gif-list'

describe('GifList', () => {
  afterEach(cleanup)

  it('should render GifList component', () => {
    // Arrange
    // Act
    const { asFragment } = render(<GifList list={[]} loading={false} />)

    // Assert
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render a list of elements when list has elements to render', () => {
    // Arrange
    const list: (TResult | TResultTrending)[] = [
      { id: 'fake id1', title: 'title1', url: '/this/is/url1' },
      { id: 'fake id2', title: 'title2', url: '/this/is/url2' },
    ]

    // Act
    const { getByTestId, getAllByTestId } = render(<GifList list={list} loading={false} />)
    const listElement = getByTestId('gif-list')
    const previews = getAllByTestId('gif-preview')

    // Assert
    expect(listElement.childElementCount).toEqual(2)
    expect(previews.length).toEqual(2)
  })
})
