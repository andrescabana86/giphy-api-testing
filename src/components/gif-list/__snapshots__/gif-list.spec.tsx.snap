// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifList should render GifList component 1`] = `
<DocumentFragment>
  <section
    class="gif-list"
    data-testid="gif-list"
  />
</DocumentFragment>
`;
